doc/developer/design/20240502_deprecate_subsources.md

```diff
@@ -0,0 +1,190 @@
+    - [storage/sources: make (sub)sources a flavor of tables #20208
+      ](https://github.com/MaterializeInc/materialize/issues/20208)
+    - [sql/storage: consider if TEXT COLUMNS, IGNORE COLUMNS belongs on
+      subsources
+      #26774](https://github.com/MaterializeInc/materialize/issues/26774)
+
+## The Problem
+
+There are two distinct notions of `SUBSOURCE`s in MZ, both of which can be
+folded neatly into other concepts.
+
+- Ingestion export subsources should be tables (as should the current primary
+  source's collection)
+- Progress subsources should become the relation returned when querying sources
+
+Doing both of these lets us remove the concept of subsources from Materialize
+entirely. It, by definition, also lets us get rid of the notion of "progress
+subsources," which is a concept that users often find confusion.
+
+## Success Criteria
+
+- Users no longer need to use the keyword `SUBSOURCE` to interact with
+  Materialize in any capacity; and we do this without deprecating any
+  functionality whatsoever.
+
+## Solution Proposal
+
+### SOURCEs should output to TABLES
+
+In the current version of Materialize, we have an impedance mismatch between
+single-output and multi-output sources.
+
+- Single-ouput sources write their data directly to the `SOURCE` itself, i.e.
+  users query the `SOURCE` object to read the data.
+- Multi-output sources write their data to `SUBSOURCES`. The primary source to
+  which they belong is an empty, unused (though still queryable) collection.
+
+I propose that the output of all sources, both single- and multi-output, should
+be to `TABLE`s (albeit special, read-only tables). Users already have an
+intuition around how `TABLE`s work, and understanding "read-only" tables is much
+simpler than doubling down on subsources––a concept we invented.
+
+#### Why not just rename `SUBSOURCE` to `TABLE`?
+
+This leaves the impedance mismatch in place where users have to `SELECT` from
+the source if they're using e.g. Kafka, but have to find the right `TABLE` to
+`SELECT` from for PG sources.
+
+### SOURCES should contain their progress information
+
+The other type of `SUBSOURCE` Materialize supports is known as the
+remap/progress collection. While most users don't access this data, it's an
+important debugging tool.
+
+I propose that the progress collection should be moved "into" the source, i.e.
+querying a source should return the what-is-today-the-progress-collection's
+data.
+
+### New API
+
+The API to support this is similar to the existing `CREATE SOURCE` API. However,
+`ALTER SOURCE...ADD SUBSOURCE` will disappear and be replaced with a flavor of
+`CREATE TABLE`.
+
+(h/t @benesch for this)
+
+***
+
+Create a Kafka source that reads from a topic named `t`:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR ALL TABLES
+```
+
+`kafka_src` is the progress relation for the source, and a table named `t` is
+automatically created in the same schema as `kafka_src`.
```

Comment:
How does this work for topics that have SQL names that are not valid as bare identifiers? These are pretty common in the Kafka world. E.g., `<project>.<topic>` or `multi-word-kebab-case-topic-names`.

I think it's safest to use the idents directly (unless they're too long), because otherwise there might be collisions, but that'll be fairly unergonomic. We may want to introduce a simple transformation scheme, like all `.` and `-` characters are transformed to `_` characters. This won't guarantee that all Kafka topics get munged to a valid bare identifier, but it'll catch the vast majority of them, and is much simpler to explain than something more complicated that gets 100% of Kafka topics.

Reply:
Everything will get "desugared" into `<some version of the topic name> AS ident_version_of_the_topic_name`, so the munging is only necessary if the user doesn't provide the table name. I think that we can convert that into a topic name like
1. Replacing `.` and `-` with `_` 
2. Stripping out any other punctuation
3. Tying to turn it into a regular `Ident`
4. If that fails, wrap it in double quotes and convert that to an `Ident`

I'm not sure if you're worried about the reference to the topic name itself, though. We don't have a strict requirement that `<some version of the topic name>` be an `Ident` (or an "unqualified" `UnresolvedItemName`); the ship hasn't sailed on what the "left-hand-side" of the reference may be. They could just be `String`s. I have a sense that we could convert them to `Ident`s through sheer force of will (you can escape basically anything), but maybe there's some incompatibility I don't see.

Reply:
> We don't have a strict requirement that `<some version of the topic name>` be an `Ident` (or an "unqualified" `UnresolvedItemName`); the ship hasn't sailed on what the "left-hand-side" of the reference may be.

Looks like there's a discussion brewing in Slack on this subject, but ah, in my head that ship had sailed! I feel like I was told at some point that in a `TABLES (upstream_name AS local_name)` clause, `upstream_name` must always be an object name according to Materialize-style rules. For upstream systems that don't have three level catalogs (e.g., MySQL doesn't have schemas), we find some way to map their catalogs to our catalogs.

----

Having let this brew in my subconscious for the past few days, I think the cleanest thing to do is not apply any automatic munging and use the Kafka name directly as an identifier if the user doesn't specify what name to use. This involves no magic and is easy to explain. It's not wholly ergonomic, but we can solve that with education—e.g., the docs could always show using the explicit `TABLES ("kafka-topic" as kafka_topic)` syntax to choose a name that's safe for use as a bare identifier.